File: frameworks/go/gorilla-mux/Implementation/parameters.go
package implementation

import (
	"net/http"
	"strconv"

	"github.com/gorilla/mux"
)

type ParametersOne struct {
	One int `json:"one"`
}

type ParametersTwo struct {
	One int `json:"one"`
	Two int `json:"two"`
}

// parametersRoutes read router captures with mux.Vars, which hands them back as strings, and
// convert them with strconv. mux tries routes in the order they are registered, and {one}
// matches the segment static as readily as a number, so the static route comes first.
func parametersRoutes(r *mux.Router, p *Payloads) {
	r.HandleFunc("/parameters/static/segment/literal", func(w http.ResponseWriter, r *http.Request) { respond(w, http.StatusOK, &p.Small) }).Methods(http.MethodGet)

	r.HandleFunc("/parameters/{one}/segment/literal", func(w http.ResponseWriter, r *http.Request) {
		one, err := strconv.Atoi(mux.Vars(r)["one"])
		if err != nil {
			refuse(w, err)
			return
		}
		respond(w, http.StatusOK, Echoed[ParametersOne]{&p.Small, ParametersOne{One: one}})
	}).Methods(http.MethodGet)

	r.HandleFunc("/parameters/{one}/with-second/{two}", func(w http.ResponseWriter, r *http.Request) {
		vars := mux.Vars(r)
		one, err := strconv.Atoi(vars["one"])
		if err != nil {
			refuse(w, err)
			return
		}
		two, err := strconv.Atoi(vars["two"])
		if err != nil {
			refuse(w, err)
			return
		}
		respond(w, http.StatusOK, Echoed[ParametersTwo]{&p.Small, ParametersTwo{One: one, Two: two}})
	}).Methods(http.MethodGet)
}

File: frameworks/rust/actix-web/Implementation/routes/parameters.rs
use actix_web::web::{self, ServiceConfig};
use serde::{Deserialize, Serialize};

use crate::Payloads;
use crate::answers::Echoed;

#[derive(Deserialize, Serialize)]
struct One {
    one: i64,
}

#[derive(Deserialize, Serialize)]
struct Two {
    one: i64,
    two: i64,
}

/// parameters: router captures, bound as integers into a struct by actix-web's Path extractor,
/// which names each field after its capture. The router tries routes in the order they were
/// registered, so the literal route comes first.
pub fn configure(cfg: &mut ServiceConfig, p: &'static Payloads) {
    cfg.route("/parameters/static/segment/literal", web::get().to(move || async move { web::Json(&p.small) }))
        .route("/parameters/{one}/segment/literal", web::get().to(move |one: web::Path<One>| async move { web::Json(Echoed::new(&p.small, one.into_inner())) }))
        .route("/parameters/{one}/with-second/{two}", web::get().to(move |two: web::Path<Two>| async move { web::Json(Echoed::new(&p.small, two.into_inner())) }));
}
